guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+    });
+  });
+});
+```
+
+### Runtime Jest configuration
+
+[Jest runs all test files in their own sandbox](https://cpojer.net/posts/building-a-javascript-testing-framework#isolate-tests-from-each-other), which ensures that any configuration options set at runtime, such as `testTimeout`, are properly reset per test. However, this behavior is not well known, and one should use alternatives if possible so as not to confuse readers by creating the illusion that global state is being modified. For instance, timeouts can be specified per test:
+
+🚫
+
+``` typescript
+jest.testTimeout(5000);
+
+describe("some-module", () => {
+  it("should do something", () => {
+    // ...
+  });
+
+  it("should do something else", () => {
+    // ...
+  });
+});
+```
+
+✅
+
+``` typescript
+describe("some-module", () => {
+  it("should do something", () => {
+    // ...
+  }, 5000);
+
+  it("should do something else", () => {
+    // ...
+  }, 5000);
+});
+```
+
+### Read more
+
+- ["Eradicating Non-Determinism in Tests"](https://martinfowler.com/articles/nonDeterminism.html#LackOfIsolation) by Martin Fowler
+
+## Avoid `beforeEach` and `afterEach`
+
+When writing tests that need to be set up or torn down in a similar way, it may be tempting to extract the setup and teardown phases to `beforeEach` and `afterEach` hooks:
+
+🚫
+
+``` typescript
+describe('TokensController', () => {
+  let controller: TokensController;
+  let onNetworkDidChangeListener: (networkState: NetworkState) => void | undefined;
+
+  beforeEach(() => {
+    onNetworkDidChange = jest.fn();
+    controller = new TokensController({
+      chainId: '0x1',
+      onNetworkDidChange: (listener) => {
+        onNetworkDidChangeListener = listener;
+      },
+    });
+  });
+
+  afterEach(() => {
+    controller.destroy();
+  });
+
+  describe('addToken', () => {
+    it('registers the given token under the default network, assuming it has not changed yet', () => {
+      controller.addToken({ symbol: 'DAI' });
+
+      expect(controller.state.tokens).toStrictEqual({
+        '0x1': {
+          symbol: 'DAI',
+        },
+      });
+    });
+
+    it('registers the given token under the current network, even after it changes', () => {
+      onNetworkDidChangeListener!({ chainId: '0x42' });
+
+      controller.addToken({ symbol: 'DAI' });
+
+      expect(controller.state.tokens).toStrictEqual({
+        '0x42': {
+          symbol: 'DAI',
+        },
+      });
+    });
+  });
+});
+```
+
+However, this strategy ends up increasing the long-term maintenance cost of tests for a couple of reasons:
+
+- It makes debugging tests difficult later. Data that is critical to the purpose of a test — i.e., that is used in the execution or verification phase — may be hidden in a `beforeEach` hook, or it may be split across `beforeEach` and the test in question. This means that in order to trace data as it moves through the test, one must spend time piecing together the full picture.
```

Comment:
Nit: What's described here sounds more like "reading" than "debugging". I have this problem not just when debugging, but also when reading a test to understand what it does, or reviewing a test in a PR.

```suggestion
- It makes tests harder to read. Data that is critical to the purpose of a test — i.e., that is used in the execution or verification phase — may be hidden in a `beforeEach` hook, or it may be split across `beforeEach` and the test in question. This means that in order to trace data as it moves through the test, one must spend time piecing together the full picture.
```

Reply:
Okay fair! Updated in [7a34dfe](https://github.com/MetaMask/contributor-docs/pull/38/commits/7a34dfe959d7e2cfa57861cd660a42e61fc42665).